File: poly_apps/laravel_main/app/Services/DeveloperHistory/Extractors/CodexExtractor.php
<?php

namespace App\Services\DeveloperHistory\Extractors;

/**
 * OpenAI Codex CLI extractor (paths confirmed via openai/codex docs).
 *
 *   <home>/.codex/sessions/rollout-<ts>-<uuid>.jsonl   per-session rollout (JSONL)
 *   (newer builds bucket by date: sessions/YYYY/MM/DD/rollout-*.jsonl)
 *   <home>/.codex/history.jsonl                        typed-prompt log
 *
 * Each rollout line is a canonical rollout item; newer builds wrap it as
 * {timestamp,type,payload}. First line is a session_meta. Sub-agents are
 * separate rollout files (no in-line sidechain flag).
 */
class CodexExtractor extends AbstractExtractor
{
    public function tool(): string
    {
        return 'codex';
    }

    public function discover(string $home, string $user): array
    {
        $root = $home . '/.codex';
        if (!is_dir($root)) {
            return [];
        }
        $out = [];
        $sessionsDir = $root . '/sessions';
        if (is_dir($sessionsDir)) {
            foreach ($this->findRollouts($sessionsDir) as $file) {
                $out[] = $this->descriptor($file);
            }
        }
        $history = $root . '/history.jsonl';
        if (is_file($history)) {
            $out[] = $this->descriptor($history);
        }
        return $out;
    }

    public function parseSource(string $path, string $user): array
    {
        if (basename($path) === 'history.jsonl') {
            $sess = $this->parseGlobalPrompts($path, $user);
            return $sess !== null ? [$sess] : [];
        }
        $sess = $this->parseRollout($path, $user);
        return $sess !== null ? [$sess] : [];
    }

    /** Recursively collect rollout-*.jsonl files (flat or date-bucketed). */
    private function findRollouts(string $dir, int $depth = 0): array
    {
        if ($depth > 6) {
            return []; // date buckets are YYYY/MM/DD; bound against symlink cycles
        }
        $found = [];
        foreach (@glob($dir . '/*') ?: [] as $path) {
            if (is_dir($path)) {
                if (is_link($path)) {
                    continue;
                }
                $found = array_merge($found, $this->findRollouts($path, $depth + 1));
            } elseif (str_ends_with($path, '.jsonl') && str_contains(basename($path), 'rollout')) {
                $found[] = $path;
            }
        }
        return $found;
    }

    private function parseRollout(string $file, string $user): ?array
    {
        $entries = $this->loadJsonl($file);
        if (empty($entries)) {
            return null;
        }

        $turns = [];
        $prompts = [];
        $sessionId = '';
        $project = '';
        $first = 0;
        $last = 0;

        foreach ($entries as $e) {
            $ts = $this->tsToEpoch($e['timestamp'] ?? null);
            if ($ts > 0) {
                $last = $ts;
                if ($first === 0) {
                    $first = $ts;
                }
            }
            $type = $e['type'] ?? ($e['record_type'] ?? '');
            $payload = $e['payload'] ?? $e;
            $ptype = $payload['type'] ?? '';

            // Only a genuine session_meta record is metadata; do NOT treat any
            // line that merely carries a 'cwd' as meta (would drop its content).
            if ($type === 'session_meta' || $ptype === 'session_meta') {
                $sessionId = $sessionId !== '' ? $sessionId : (string) ($payload['id'] ?? '');
                $project = $project !== '' ? $project : (string) ($payload['cwd'] ?? '');
                continue;
            }
            if ($project === '' && isset($payload['cwd'])) {
                $project = (string) $payload['cwd'];
            }
            if ($ptype === 'message') {
                $role = (string) ($payload['role'] ?? 'assistant');
                $text = $this->extractResponseText($payload['content'] ?? []);
                if (trim($text) === '') {
                    continue;
                }
                if ($role === 'user') {
                    $prompts[] = ['ts' => $ts, 'text' => $this->truncate($text)];
                    $turns[] = $this->turn($ts, 'user', $text);
                } else {
                    $turns[] = $this->turn($ts, 'assistant', $text);
                }
            } elseif ($ptype === 'function_call') {
                $args = json_encode($payload['arguments'] ?? null, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
                $turns[] = $this->turn($ts, 'tool_use', (string) $args, false, null, (string) ($payload['name'] ?? '?'));
            } elseif ($ptype === 'function_call_output') {
                $turns[] = $this->turn($ts, 'tool_result', (string) ($payload['output'] ?? ''));
            }

            if (count($turns) > self::MAX_TURNS) {
                break;
            }
        }

        if ($sessionId === '') {
            $sessionId = pathinfo($file, PATHINFO_FILENAME);
        }
        if (empty($turns)) {
            return null;
        }

        return $this->session('codex', $user, $sessionId, [
            'project' => $project,
            'firstTs' => $first,
            'lastTs' => $last,
            'source' => $file,
            'prompts' => $prompts,
            'turns' => $turns,
        ]);
    }

    /** Responses-API content[] -> text (input_text/output_text blocks). */
    private function extractResponseText($content): string
    {
        if (is_string($content)) {
            return $content;
        }
        if (!is_array($content)) {
            return '';
        }
        $parts = [];
        foreach ($content as $b) {
            if (is_array($b) && isset($b['text'])) {
                $parts[] = (string) $b['text'];
            } elseif (is_string($b)) {
                $parts[] = $b;
            }
        }
        return implode("\n", $parts);
    }

    private function parseGlobalPrompts(string $src, string $user): ?array
    {
        $rows = $this->loadJsonl($src);
        if (empty($rows)) {
            return null;
        }
        $prompts = [];
        $turns = [];
        $first = 0;
        $last = 0;
        foreach ($rows as $d) {
            $text = trim((string) ($d['text'] ?? ($d['display'] ?? '')));
            if ($text === '') {
                continue;
            }
            $ts = $this->tsToEpoch($d['ts'] ?? ($d['timestamp'] ?? null));
            if ($ts > 0) {
                $last = $ts;
                if ($first === 0) {
                    $first = $ts;
                }
            }
            $prompts[] = ['ts' => $ts, 'text' => $this->truncate($text)];
            $turns[] = $this->turn($ts, 'user', $text);
        }
        if (empty($prompts)) {
            return null;
        }

        return $this->session('codex', $user, 'global-typed-prompts', [
            'project' => '(all projects)',
            'title' => 'Typed prompts (global history.jsonl)',
            'firstTs' => $first,
            'lastTs' => $last,
            'source' => $src,
            'prompts' => $prompts,
            'turns' => $turns,
        ]);
    }
}

File: pycore/pyctl/agent_history/codex_extractor.py
# -*- coding: utf-8 -*-
"""OpenAI Codex CLI extractor — mirrors Laravel CodexExtractor."""

from __future__ import annotations

import json
import os
from glob import glob
from typing import Any, Dict, List, Optional

from pycore.pyctl.agent_history.base_extractor import BaseExtractor, MAX_TURNS


class CodexExtractor(BaseExtractor):
    def tool(self) -> str:
        return "codex"

    def discover(self, home: str, user: str) -> List[Dict[str, Any]]:
        root = os.path.join(home, ".codex")
        if not os.path.isdir(root):
            return []
        out: List[Dict[str, Any]] = []
        sessions_dir = os.path.join(root, "sessions")
        if os.path.isdir(sessions_dir):
            for file in self._find_rollouts(sessions_dir):
                out.append(self.descriptor(file))
        history = os.path.join(root, "history.jsonl")
        if os.path.isfile(history):
            out.append(self.descriptor(history))
        return out

    def _find_rollouts(self, directory: str, depth: int = 0) -> List[str]:
        if depth > 6:
            return []
        found: List[str] = []
        for path in glob(os.path.join(directory, "*")):
            if os.path.isdir(path):
                if os.path.islink(path):
                    continue
                found.extend(self._find_rollouts(path, depth + 1))
            elif path.endswith(".jsonl") and "rollout" in os.path.basename(path):
                found.append(path)
        return found

    def parse_source(self, path: str, user: str) -> List[Dict[str, Any]]:
        if os.path.basename(path) == "history.jsonl":
            sess = self._parse_global_prompts(path, user)
            return [sess] if sess else []
        sess = self._parse_rollout(path, user)
        return [sess] if sess else []

    def _parse_rollout(self, file: str, user: str) -> Optional[Dict[str, Any]]:
        entries = self.load_jsonl(file)
        if not entries:
            return None
        turns: List[Dict[str, Any]] = []
        prompts: List[Dict[str, Any]] = []
        session_id = ""
        project = ""
        first = 0
        last = 0

        for e in entries:
            ts = self.ts_to_epoch(e.get("timestamp"))
            if ts > 0:
                last = ts
                if first == 0:
                    first = ts
            etype = e.get("type") or e.get("record_type") or ""
            payload = e.get("payload") or e
            ptype = payload.get("type", "")

            if etype == "session_meta" or ptype == "session_meta":
                session_id = session_id or str(payload.get("id") or "")
                project = project or str(payload.get("cwd") or "")
                continue
            if not project and payload.get("cwd"):
                project = str(payload["cwd"])

            if ptype == "message":
                role = str(payload.get("role") or "assistant")
                text = self._extract_response_text(payload.get("content") or [])
                if not text.strip():
                    continue
                if role == "user":
                    prompts.append({"ts": ts, "text": self.truncate(text)})
                    turns.append(self.turn(ts, "user", text))
                else:
                    turns.append(self.turn(ts, "assistant", text))
            elif ptype == "function_call":
                args = json.dumps(payload.get("arguments"), ensure_ascii=False)
                turns.append(self.turn(ts, "tool_use", args, False, None, str(payload.get("name") or "?")))
            elif ptype == "function_call_output":
                turns.append(self.turn(ts, "tool_result", str(payload.get("output") or "")))

            if len(turns) > MAX_TURNS:
                break

        if not session_id:
            session_id = os.path.splitext(os.path.basename(file))[0]
        if not turns:
            return None

        return self.session("codex", user, session_id, {
            "project": project,
            "firstTs": first,
            "lastTs": last,
            "source": file,
            "prompts": prompts,
            "turns": turns,
        })

    def _extract_response_text(self, content: Any) -> str:
        if isinstance(content, str):
            return content
        if not isinstance(content, list):
            return ""
        parts: List[str] = []
        for b in content:
            if isinstance(b, dict) and "text" in b:
                parts.append(str(b["text"]))
            elif isinstance(b, str):
                parts.append(b)
        return "\n".join(parts)

    def _parse_global_prompts(self, src: str, user: str) -> Optional[Dict[str, Any]]:
        rows = self.load_jsonl(src)
        if not rows:
            return None
        prompts: List[Dict[str, Any]] = []
        turns: List[Dict[str, Any]] = []
        first = 0
        last = 0
        for d in rows:
            text = str(d.get("text") or d.get("display") or "").strip()
            if not text:
                continue
            ts = self.ts_to_epoch(d.get("ts") or d.get("timestamp"))
            if ts > 0:
                last = ts
                if first == 0:
                    first = ts
            prompts.append({"ts": ts, "text": self.truncate(text)})
            turns.append(self.turn(ts, "user", text))
        if not prompts:
            return None
        return self.session("codex", user, "global-typed-prompts", {
            "project": "(all projects)",
            "title": "Typed prompts (global history.jsonl)",
            "firstTs": first,
            "lastTs": last,
            "source": src,
            "prompts": prompts,
            "turns": turns,
        })
